python/ray/services.py

```diff
@@ -611,21 +586,91 @@ def start_redis(node_ip_address,
     return redis_address, redis_shards, processes
 
 
-def _start_redis_instance(node_ip_address="127.0.0.1",
+def _start_redis(node_ip_address,
+                 use_credis,
+                 is_primary,
+                 port=None,
+                 password=None,
+                 redis_max_clients=None,
+                 redis_max_memory=None,
+                 stdout_file=None,
+                 stderr_file=None):
+    """Start a single Redis server.
+
+    Args:
+        node_ip_address: The IP address of the current node. This is only used
+            for recording the log filenames in Redis.
+        use_credis: If True, additionally load the chain-replicated libraries
+            into the redis servers.
+        is_primary (bool): Is this redis shard the primary shard?
+        port (int): If provided, start a Redis server with this port.
+        password (str): Prevents external clients without the password
+            from connecting to Redis if provided.
+        redis_max_clients: If this is provided, Ray will attempt to configure
+            Redis with this maxclients number.
+        redis_max_memory: The max amount of memory (in bytes) to allow redis
+            to use, or None for no limit. Once the limit is exceeded, redis
+            will start LRU eviction of entries.
+        stdout_file: A file handle opened for writing to redirect stdout to. If
+            no redirection should happen, then this should be None.
+        stderr_file: A file handle opened for writing to redirect stderr to. If
+            no redirection should happen, then this should be None.
+
+    Returns:
+        A tuple of the port used by Redis and ProcessInfo for the process that
+            was started. If a port is passed in, then the returned port value
+            is the same.
+
+    Raises:
+        Exception: An exception is raised if Redis could not be started.
+    """
+    if use_credis:
+        redis_executable = CREDIS_EXECUTABLE
+        # It is important to load the credis module BEFORE the ray module,
+        # as the latter contains an extern declaration that the former
+        # supplies.
+        if is_primary:
```

Comment:
I see that this is similar to the behavior before, but the behavior before seems incorrect. I think only the non-primary Redis shards should use `credis`, and they should all use the `CREDIS_MASTER_MODULE`. Let's get rid of this `is_primary` flag.

Reply:
But when you are using credis, you have to config the chain state (in the original code):

```python
# Configure the chain state.
        primary_redis_client.execute_command("MASTER.ADD", node_ip_address,
                                             redis_shard_port)
        shard_client.execute_command("MEMBER.CONNECT_TO_MASTER",
                                     node_ip_address, port)
```

So you have to use `CREDIS_MASTER_MODULE` for primary credis, and `CREDIS_MEMBER_MODULE` for non-primary credis.

Reply:
The way it is intended to work is the following.

```
PRIMARY_SHARD

SHARD_1 (master replica) -> SHARD_1 (member replica) -> SHARD_1 (member replica)

SHARD_2 (master replica) -> SHARD_2 (member replica) -> SHARD_2 (member replica)

...
```

In this PR (and the codebase), it looks like

```
PRIMARY_SHARD (master) -> SHARD_1 (member replica) -> SHARD_2 (member replica) -> ...
```

However, the first way is correct.